components/sha2017/sha2017_ota.c

```diff
@@ -72,417 +63,503 @@ static const char *REQUEST = "GET " BADGE_OTA_WEB_PATH " HTTP/1.0\r\n"
-    }
-  }
-
-  mbedtls_ssl_close_notify(&ssl);
-
-  ESP_LOGI(TAG, "Total Write binary data length : %d", binary_file_length);
-
-  if (esp_ota_end(update_handle) != ESP_OK) {
-    ESP_LOGE(TAG, "esp_ota_end failed!");
-    task_fatal_error();
-  }
-
-  if (binary_file_length != content_length) {
-    ESP_LOGE(TAG, "Firmware size differs, expected: %d\nReceived: %d",
-             content_length, binary_file_length);
-    task_fatal_error();
-  }
-
-  /* FIXME: we should really add code here which verifies the integrity of the
-   * new OTA partition.
-   */
-
-   show_percentage("Rebooting the badge", 0, false);
-
-  err = esp_ota_set_boot_partition(update_partition);
-  if (err != ESP_OK) {
-    ESP_LOGE(TAG, "esp_ota_set_boot_partition failed! err=0x%x", err);
-    task_fatal_error();
-  }
-  ESP_LOGI(TAG, "Prepare to restart system!");
-  esp_restart();
-  return;
+	ret = mbedtls_ssl_setup(&ssl, &conf);
+	if (ret != 0) {
+		ESP_LOGE(TAG, "mbedtls_ssl_setup returned -0x%x\n\n", -ret);
+		task_fatal_error();
+	}
+
+	/* Wait for the callback to set the CONNECTED_BIT in the
+	   event group.
+	 */
+	xEventGroupWaitBits(wifi_event_group, CONNECTED_BIT, false, true,
+			portMAX_DELAY);
+	ESP_LOGI(TAG, "Connected to AP");
+
+	mbedtls_net_init(&server_fd);
+
+	show_percentage("Handshaking server", 0, false);
+
+	ESP_LOGI(TAG, "Connecting to %s:%s...", BADGE_OTA_WEB_SERVER,
+			BADGE_OTA_WEB_PORT);
+
+	ret = mbedtls_net_connect(&server_fd, BADGE_OTA_WEB_SERVER, BADGE_OTA_WEB_PORT, MBEDTLS_NET_PROTO_TCP);
+	if (ret != 0) {
+		ESP_LOGE(TAG, "mbedtls_net_connect returned -%x", -ret);
+		task_fatal_error();
+	}
+
+	ESP_LOGI(TAG, "Connected.");
+
+	mbedtls_ssl_set_bio(&ssl, &server_fd, mbedtls_net_send, mbedtls_net_recv,
+			NULL);
+
+	ESP_LOGI(TAG, "Performing the SSL/TLS handshake...");
+
+	ret = mbedtls_ssl_handshake_(&ssl);
+	if (ret != 0) {
+		ESP_LOGE(TAG, "mbedtls_ssl_handshake returned -0x%x", -ret);
+		task_fatal_error();
+	}
+
+	ESP_LOGI(TAG, "Verifying peer X.509 certificate...");
+
+	/* NOTE: Afaik, the mbedtls_ssl_get_verify_result() always returns 0 if
+	 *       MBEDTLS_SSL_VERIFY_REQUIRED is used.
+	 */
+	ret = mbedtls_ssl_get_verify_result(&ssl);
+	if (ret != 0) {
+		/* In real life, we probably want to close connection if ret != 0 */
+		ESP_LOGW(TAG, "Failed to verify peer certificate!");
+		task_fatal_error();
+	}
+
+	ESP_LOGI(TAG, "Certificate verified.");
+
+	ESP_LOGI(TAG, "Sending HTTP request for %s", BADGE_OTA_WEB_PATH);
+
+	ret = mbedtls_ssl_write_(&ssl, (const unsigned char *) REQUEST, strlen(REQUEST));
+	if (ret <= 0) {
+		ESP_LOGE(TAG, "mbedtls_ssl_write returned -0x%x", -ret);
+		task_fatal_error();
+	}
+	ESP_LOGI(TAG, "%d bytes written", strlen(REQUEST));
+
+	show_percentage("Starting OTA update", 0, false);
+
+	/* read until we have received the status line */
+	ESP_LOGI(TAG, "Reading HTTP response status line.");
+	uint8_t *crlf;
+	while ((crlf = index_crlf(buffer, buffer_len)) == NULL) {
```

Comment:
Could fix this with code duplication (which results in slower code and risking forgetting the 2nd call to index_crlf()) or by splitting this in an ugly 'while (1) { crlf = index(..); if (crlf != NULL) break; ..'